fdbserver/storageserver.actor.cpp

```diff
@@ -4906,6 +4967,421 @@ ACTOR Future<Void> validateRangeAgainstServers(StorageServer* data,
+	wait(data->serveAuditStorageParallelismLock.take(TaskPriority::DefaultYield));
+	state FlowLock::Releaser holder(data->serveAuditStorageParallelismLock);
+	state Version grv;
+	state Version readAtVersion;
+	state Version versionBeforeWait;
+	state std::vector<std::pair<Version, KeyRangeRef>> shardAssignments;
+	state AuditStorageState res(req.id, data->thisServerID, req.getType());
+	state std::vector<std::string> errors;
+	state Transaction tr(data->cx);
+	// We want to find out any mismatch between ownRangesSeenByServerKey and ownRangesLocalView
+	state std::vector<KeyRange> ownRangesSeenByServerKey;
+	state std::vector<KeyRange> ownRangesLocalView;
+	try {
+		if (data->version.get() == 0) {
+			TraceEvent(SevInfo, "AuditStorageShardStorageServerShardFailed", data->thisServerID)
+			    .detail("Reason", "Version is not valid")
+			    .detail("LatestVersion", data->version.get());
+			req.reply.sendError(audit_storage_failed()); // we cannot run this aduit for this time
+			return Void();
+		}
+
+		// Read local shard info and serverKeys
+		tr.setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+		wait(store(grv, tr.getReadVersion()));
+		// When selecting which version to read, we should read at a new version
+		// As a result, any fall behind SS can eventually read the data
+		if (grv <= data->version.get()) {
+			readAtVersion = data->version.get();
+			ownRangesLocalView = getServerShardsByReadingShardInfo(data, readAtVersion);
+			tr.reset();
+			tr.setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+			tr.setVersion(readAtVersion); // this can be a future version
+			// getServerShardsByReadingServerKeys will intenally retry until get result
+			wait(store(ownRangesSeenByServerKey, getServerShardsByReadingServerKeys(data, &tr)));
+		} else {
+			// read at grv version
+			readAtVersion = grv;
+			wait(store(ownRangesSeenByServerKey, getServerShardsByReadingServerKeys(data, &tr)));
+			versionBeforeWait = data->version.get();
+			// open trace
+			data->registerAuditsForShardAssignmentHistoryCollection(req.id);
+			wait(data->version.whenAtLeast(grv));
+			// close trace
+			data->unregisterAuditsForShardAssignmentHistoryCollection(req.id);
+			ownRangesLocalView = getServerShardsByReadingShardInfo(data, data->version.get());
+			// check any serverKey update between grv and data->version.get()
+			shardAssignments = data->getAssignmentHistory(grv, data->version.get());
+			TraceEvent(SevVerbose, "AuditStorageShardStorageServerShardGetHistory", data->thisServerID)
+			    .detail("ReadAtVersion", readAtVersion)
+			    .detail("VersionBeforeWait", versionBeforeWait)
+			    .detail("VersionAfterWait", data->version.get())
+			    .detail("ShardAssignmentsCount", shardAssignments.size())
+			    .detail("GHistory", data->printAssignedShardHistory());
+			// Ideally, we revert ownRangesLocalView changes from data->version.get() to grv
+			// Currently, we return failed if any update collected
+			if (shardAssignments.size() != 0) {
```

Comment:
We count only the changes between `readAtVersion` and the current version, right? do we need to store `readAtVersion` somewhere, and only count changes after that version?

Reply:
Yes. Current, we track the updates between the old data->version.get() and the new data->version.get(). We actually only use the updates between readAtVersion and the new data->version.get(). Let me improve it.